tensorflow_probability/python/mcmc/transformed_kernel.py

```diff
@@ -110,6 +106,33 @@ def transformed_log_prob_fn(*state_parts):
   return transformed_log_prob_fn
 
 
+def target_log_prob_getter(kernel):
+  kernel_stack = [kernel]
+  while 'target_log_prob_fn' not in kernel.parameters:
+    if 'inner_kernel' not in kernel.parameters:
+      raise ValueError('"None of the nested `inner_kernel`s contains a '
+                       '`target_log_prob_fn`."')
+    kernel = kernel.inner_kernel
+    kernel_stack.append(kernel)
+  return kernel.parameters['target_log_prob_fn'], kernel_stack
+
+
+def target_log_prob_setter(kernel_stack, new_target_log_prob):
```

Comment:
This is an example of API simplicity being in tension with performance (and to some extent with implementation simplicity).  This function will be much simpler to understand and use if it accepts the single top-level kernel and the new log_prob function, and computes the kernel stack on its own.  As it is, though, the user has to know where to get the kernel stack from in order to use this, and their code has to have an extra communication path (from the getter to the setter).  This API also hurts the use-case of replacing the target log_prob unconditionally, without reading the target that was there before (not that such a use-case is very common).

Reply:
Good pointer; I hadn't even thought about it. Edited for API simplicity.